### PR TITLE
QALI-470 Fix the issue about campaign aborting

### PR DESCRIPTION
**Pull request by joval2**

### CODE REVIEW

File: head/src/main/kotlin/io/qalipsis/core/head/campaign/PersistentCampaignService.kt [hunk 1]
```diff
@@ -156,7 +157,12 @@ internal class PersistentCampaignService(
     @LogInput
     override suspend fun abort(tenant: String, aborter: String?, campaignKey: String) {
         campaignRepository.findByTenantAndKey(tenant, campaignKey)?.let { campaign ->
-            campaignRepository.update(campaign.copy(aborter = aborter?.let { userRepository.findIdByUsername(it) }))
+            campaignRepository.update(
+                campaign.copy(
+                    aborter = aborter?.let { userRepository.findIdByUsername(it) },
+                    result = ExecutionStatus.ABORTED
```
Comment (ericjesse) [inline head/src/main/kotlin/io/qalipsis/core/head/campaign/PersistentCampaignService.kt:163]: Only if the abortion is hard.

File: head/src/main/kotlin/io/qalipsis/core/head/campaign/PersistentCampaignService.kt [hunk 2]
```diff
@@ -42,6 +42,7 @@ import io.qalipsis.core.head.utils.SortingUtil
 import io.qalipsis.core.head.utils.SqlFilterUtils.formatsFilters
 import jakarta.inject.Singleton
 import java.time.Instant
+import kotlin.math.E
```
Comment (ericjesse) [inline head/src/main/kotlin/io/qalipsis/core/head/campaign/PersistentCampaignService.kt:45]: ???

**Review by ericjesse — CHANGES_REQUESTED** [review]

File: head/src/main/kotlin/io/qalipsis/core/head/campaign/PersistentCampaignService.kt (hunk 2 above)
Comment (joval2) [reply head/src/main/kotlin/io/qalipsis/core/head/campaign/PersistentCampaignService.kt:45]: It was a mistake, I don't know what happened

Comment (joval2) [reply head/src/main/kotlin/io/qalipsis/core/head/campaign/PersistentCampaignService.kt:45]: It was a mistake, I don't know how it happened.

File: head/src/test/kotlin/io/qalipsis/core/head/campaign/states/FactoryAssignmentStateTest.kt [hunk 3]
```diff
@@ -272,9 +273,12 @@ internal class FactoryAssignmentStateTest : AbstractStateTest() {
         val newState = state.abort(AbortRunningCampaign())
 
         // then
-        assertThat(newState).isInstanceOf(AbortingState::class).all {
+        assertThat(newState).isInstanceOf(DisabledState::class).all {
```
Comment (ericjesse) [inline head/src/test/kotlin/io/qalipsis/core/head/campaign/states/FactoryAssignmentStateTest.kt:276]: The tests for the new cases that return the abort state with all the factories or just part of them are missing.
- same for all the classes -

File: head/src/main/kotlin/io/qalipsis/core/head/jdbc/repository/FactoryRepository.kt [hunk 4]
```diff
@@ -91,6 +92,31 @@ internal interface FactoryRepository : CoroutineCrudRepository<FactoryEntity, Lo
...
+     * Checks factories' health in a [tenant]. In factory_state table,
+     * we only select the row corresponding to the more recent health_timestamp.
+     *
+     * @param tenant the tenant's reference
+     * @param factoryNodeIds the collection of factories to check
+     */
+    @Query(
+        """SELECT factory_max_timestamp."node_id", 
+                factory_state."health_timestamp" as "timestamp", 
+                factory_state."state"
```
Comment (ericjesse) [inline head/src/main/kotlin/io/qalipsis/core/head/jdbc/repository/FactoryRepository.kt:105]: When the latest health for a factory is "HEALTHY" but its health_timestamp is older than 2 minutes, it should be considered as UNHEALTHY. 
Please create a proper test for it.

**Review by ericjesse — CHANGES_REQUESTED** [review]

File: head/src/main/kotlin/io/qalipsis/core/head/campaign/states/FactoryAssignmentState.kt [hunk 5]
```diff
@@ -91,7 +92,16 @@ internal open class FactoryAssignmentState(
     }
 
     override suspend fun abort(abortConfiguration: AbortRunningCampaign): CampaignExecutionState<CampaignExecutionContext> {
```
[the remark this answers is not in the document]
Comment (joval2) [reply head/src/main/kotlin/io/qalipsis/core/head/campaign/states/FactoryAssignmentState.kt:94]: I think it is better. That way, if we need to make any changes, we can do so in one place.

File: head/src/main/kotlin/io/qalipsis/core/head/campaign/states/FactoryAssignmentState.kt [hunk 6]
```diff
@@ -91,7 +92,16 @@ internal open class FactoryAssignmentState(
...
-        return AbortingState(campaign, abortConfiguration, "The campaign was aborted")
+        val healthyFactories = context.factoryService.checkFactoriesHealth(campaign.tenant, campaign.factories.keys)
+            .filter { it.state == Heartbeat.State.IDLE }.map { it.nodeId }.toSet()
+        return if (healthyFactories.isEmpty()) {
+            // If no factory is healthy return DisabledState
+            DisabledState(campaign, false)
+        } else {
+            // Else only keep healthy factories and return AbortingState
+            campaign.factories.keys.removeIf { it !in healthyFactories }
```
[the remark this answers is not in the document]
Comment (joval2) [reply head/src/main/kotlin/io/qalipsis/core/head/campaign/states/FactoryAssignmentState.kt:102]: It really update the `campaign.factories`.

File: head/src/main/kotlin/io/qalipsis/core/head/campaign/states/FactoryAssignmentState.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (joval2) [reply head/src/main/kotlin/io/qalipsis/core/head/campaign/states/FactoryAssignmentState.kt:94]: But the problem in this abstract class is that, we do not have enough (`RunningCampaign`) data to implement this method

File: head/src/main/kotlin/io/qalipsis/core/head/redis/campaign/RedisMinionsAssignmentState.kt [hunk 7]
```diff
@@ -73,6 +75,15 @@ internal class RedisMinionsAssignmentState(
     }
 
     override suspend fun abort(abortConfiguration: AbortRunningCampaign): CampaignExecutionState<CampaignExecutionContext> {
```
[the remark this answers is not in the document]
Comment (joval2) [reply head/src/main/kotlin/io/qalipsis/core/head/redis/campaign/RedisMinionsAssignmentState.kt:77]: I don't understand this comment very well. Where does the `createAbortingState` come from ? Apart from the `abort` method, I see that the other methods of the `redisX` and `non-RedisX` classes are also different.

File: head/src/test/kotlin/io/qalipsis/core/head/campaign/states/FactoryAssignmentStateTest.kt [hunk 8]
```diff
@@ -278,7 +281,109 @@ internal class FactoryAssignmentStateTest : AbstractStateTest() {
...
-            factoryService.checkFactoriesHealth(refEq("my-tenant"), setOf())
+            factoryService.getFactoriesHealth(refEq("my-tenant"), setOf())
+        }
+        confirmVerified(factoryService, campaignReportStateKeeper)
+    }
+
+    @Test
+    fun `should return an aborting state`() = testDispatcherProvider.runTest {
+        // given
+        val state = FactoryAssignmentState(campaign, mockk(), mockk())
```
Comment (ericjesse) [inline head/src/test/kotlin/io/qalipsis/core/head/campaign/states/FactoryAssignmentStateTest.kt:292]:
```suggestion
       every { campaign.factories.keys } returns mutableSetOf("node-1", "node-2", "node-3")
        coEvery { factoryService.getFactoriesHealth("my-tenant", mutableSetOf("node-1", "node-2", "node-3")) }      returns emptySet()
        val state = FactoryAssignmentState(campaign, mockk(), mockk())
```

Apply this change in all the test classes.

**Review by ericjesse — CHANGES_REQUESTED** [review]

File: head/src/test/kotlin/io/qalipsis/core/head/campaign/states/FactoryAssignmentStateTest.kt (hunk 8 above)
Comment (ericjesse) [reply head/src/test/kotlin/io/qalipsis/core/head/campaign/states/FactoryAssignmentStateTest.kt:292]: Do not rename the function, I had not seen that you had create the use case as well for the partially unhealthy factories. Just add the mocks in that function.

Comment (ericjesse) [reply head/src/test/kotlin/io/qalipsis/core/head/campaign/states/FactoryAssignmentStateTest.kt:292]: Sorry, the mocks should actually be in the function `should return a disabled state...`

**Review by ericjesse — APPROVED** [review]

**Pull request merged**